File: python-refactor/Code/pre_03_station/JP_station_03_stn_code.py
### Package Import
import sys
import os
base_dir = os.environ['GEMS_HOME']
project_path = os.path.join(base_dir, 'python-refactor')
sys.path.insert(0, project_path)
from Code.utils import matlab

import copy
import numpy as np
import pandas as pd
import glob
import h5py

### Setting path
data_base_dir = os.path.join('/data2', 'sehyun', 'Data')
path_station = os.path.join(data_base_dir, 'Preprocessed_raw', 'Station') 
path_stn_jp = os.path.join(path_station, 'Station_JP')

stn_info = pd.read_csv(os.path.join(path_stn_jp, 'jp_stn_code_lonlat_period_filtered_yyyymmdd.csv'))
stn_info = stn_info.values
# scode1, scode2, lon, lat, op_start, op_
scode_unq = np.unique(stn_info[:,0])
del stn_info 

header_ndata = np.array(['doy','yr','mon','day','KST','SO2','CO','OX','NO2','PM10','PM25',
    'NO','NOX','NMHC','CH4','THC','CO2','scode'],
                        dtype=h5py.string_dtype(encoding='utf-8'))

YEARS = [2016]
for yr in YEARS: #:2016 #2009:2016
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnSO2_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnSO2 = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnNO_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnNO = df[mat['header']].values
    
    mat  = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnNO2_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnNO2 = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnNOX_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnNOX = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnCO_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnCO = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnOX_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnOX = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnNMHC_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnNMHC = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnCH4_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnCH4 = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnTHC_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnTHC = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnSPM_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnSPM = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnPM25_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnPM25 = df[mat['header']].values
    
    mat = matlab.loadmat(os.path.join(path_stn_jp, 'byPollutant/',f'JP_stnCO2_{yr}.mat'))
    cols = list(mat.keys())
    cols.remove('header')
    df = pd.DataFrame.from_dict(mat)
    stnCO2 = df[mat['header']].values
    
    if yr%4==0: days = 366
    else: days = 365
    
    a_doy,a_KST,a_scode = np.meshgrid(np.asarray(range(1, days+1)), np.asarray(range(1, 24+1)), scode_unq)
    aa = np.hstack([a_doy.ravel(order='F'),a_KST.ravel(order='F'),a_scode.ravel(order='F')])
    
    doy000 = matlab.datenum(f'{yr}00000')   
    mm = [d[4:6] for d in matlab.datestr(doy000+aa[:,0])] # datestr: yyyymmdd
    dd = [d[6:] for d in matlab.datestr(doy000+aa[:,0])]    
    bb = np.full((aa.shape[0], 12), np.nan)
    nanidx = None
    for k in range(np.max(aa.shape)):
        tStart = time.time()
        aSO1 = stnSO1[(stnSO1[:,0]==aa[k,0]) & (stnSO1[:,6]==aa[k,1]) & (stnSO1[:,4]==aa[k,2]),7]
        aNO = stnNO[(stnNO[:,0]==aa[k,0]) & (stnNO[:,6]==aa[k,1]) & (stnNO[:,4]==aa[k,2]),7]
        aNO1 = stnNO1[(stnNO1[:,0]==aa[k,0]) & (stnNO1[:,6]==aa[k,1]) & (stnNO1[:,4]==aa[k,2]),7]
        aNOX = stnNOX[(stnNOX[:,0]==aa[k,0]) & (stnNOX[:,6]==aa[k,1]) & (stnNOX[:,4]==aa[k,2]),7]
        aCO = stnCO[(stnCO[:,0]==aa[k,0]) & (stnCO[:,6]==aa[k,1]) & (stnCO[:,4]==aa[k,2]),7]
        aOX = stnOX[(stnOX[:,0]==aa[k,0]) & (stnOX[:,6]==aa[k,1]) & (stnOX[:,4]==aa[k,2]),7]
        aNMHC = stnNMHC[(stnNMHC[:,0]==aa[k,0]) & (stnNMHC[:,6]==aa[k,1]) & (stnNMHC[:,4]==aa[k,2]),7]
        aCH4 = stnCH4[(stnCH4[:,0]==aa[k,0]) & (stnCH4[:,6]==aa[k,1]) & (stnCH4[:,4]==aa[k,2]),7]
        aTHC = stnTHC[(stnTHC[:,0]==aa[k,0]) & (stnTHC[:,6]==aa[k,1]) & (stnTHC[:,4]==aa[k,2]),7]
        aSPM = stnSPM[(stnSPM[:,0]==aa[k,0]) & (stnSPM[:,6]==aa[k,1]) & (stnSPM[:,4]==aa[k,2]),7]
        aPM14 = stnPM14[(stnPM14[:,0]==aa[k,0]) & (stnPM14[:,6]==aa[k,1]) & (stnPM14[:,4]==aa[k,2]),7]
        aCO1 = stnCO1[(stnCO1[:,0]==aa[k,0]) & (stnCO1[:,6]==aa[k,1]) & (stnCO1[:,4]==aa[k,2]),7]
        if len(aSO2)!=0: bb[k,0]=aSO2 
        if len(aCO)!=0:  bb[k,1]=aCO 
        if len(aOX)!=0:  bb[k,2]=aOX 
        if len(aNO2)!=0:  bb[k,3]=aNO2 
        if len(aSPM)!=0:  bb[k,4]=aSPM 
        if len(aSPM)!=0:  bb[k,5]=aPM25 
        if len(aNO)!=0:  bb[k,6]=aNO 
        if len(aNOX)!=0:  bb[k,7]=aNOX 
        if len(aSaNMHCO2)!=0:  bb[k,8]=aNMHC 
        if len(aCH4)!=0:  bb[k,9]=aCH4 
        if len(aTHC)!=0:  bb[k,10]=aTHC 
        if len(aCO2)!=0:  bb[k,11]=aCO2 
        
        bb_temp = bb[k,:]
        bb_temp[bb_temp>=9997]= np.nan
        bb[k,:]=bb_temp
        
        nanidx_temp = np.sum(np.isnan(bb_temp))
        if nanidx_temp == 12:
            if nanidx is None:
                nanidx = k
            else:
                nanidx = np.vstack([nanidx, k])
        
        tElapsed = time.time() - tStart
        print (f'{yr}_{aa[k,0]:03d}_{aa[k,1]:02d}---{tElapsed} sec')
    
    ndata = np.hstack([aa[:,0], np.multiply(yr, np.ones((aa.shape[0], 1))), mm, dd, aa[:,1], bb, aa[:,2]])
    ndata = np.delete(ndata, nanidx, axis=0)
    matlab.savemat(os.path.join(path_data,'Station/Station_JP/stn_code_data', f'stn_code_data_all_{yr}.mat'),
                   {'ndata':ndata, 'header_ndata':header_ndata})
    print (yr)

File: python-refactor/Code/pre_03_station/JP_station_03_stn_code.m
clear all; close all; clc

tic

% path_data = '//10.72.26.56/irisnas5/Data/';
% path = '//10.72.26.56/irisnas5/GEMS/EA_GOCI6km/';

path_data = '/share/irisnas5/Data/';
% path = '/share/irisnas5/GEMS/EA_GOCI6km/';
%
% addpath(genpath('/share/irisnas5/Data/matlab_func/'))

stn_info = csvread([path_data,'Station/Station_JP/jp_stn_code_lonlat_period_filtered_yyyymmdd.csv'],1);
% scode1, scode2, lon, lat, op_start, op_end

scode_unq = unique(stn_info(:,1));
clearvars stn_info

header_ndata = {'doy','yr','mon','day','KST','SO2','CO','OX','NO2','PM10','PM25',...
    'NO','NOX','NMHC','CH4','THC','CO2','scode'};

for yr=2015 %:2016 %2009:2016
    load([path_data,'Station/Station_JP/byPollutant/JP_stnSO2_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnNO_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnNO2_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnNOX_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnCO_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnOX_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnNMHC_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnCH4_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnTHC_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnSPM_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnPM25_',num2str(yr)])
    load([path_data,'Station/Station_JP/byPollutant/JP_stnCO2_',num2str(yr)])
    
    stnSO2 = table2array(stnSO2_tbl);
    stnNO = table2array(stnNO_tbl);
    stnNO2 = table2array(stnNO2_tbl);
    stnNOX = table2array(stnNOX_tbl);
    stnCO = table2array(stnCO_tbl);
    stnOX = table2array(stnOX_tbl);
    stnNMHC = table2array(stnNMHC_tbl);
    stnCH4 = table2array(stnCH4_tbl);
    stnTHC = table2array(stnTHC_tbl);
    stnSPM = table2array(stnSPM_tbl);
    stnPM25 = table2array(stnPM25_tbl);
    stnCO2 = table2array(stnCO2_tbl);
    
    clearvars *_tbl
    
    if mod(yr,4)==0
        days = 366;
    else
        days = 365;
    end
    
    [a_doy,a_KST,a_scode] = meshgrid(1:days,1:24,scode_unq);
    aa = [a_doy(:),a_KST(:),a_scode(:)];
    
    doy000 = datenum([yr,0,0,0,0,0]);
    mm = str2num(datestr(doy000+aa(:,1),'mm'));
    dd = str2num(datestr(doy000+aa(:,1),'dd'));
    
%     clearvars a_doy a_KST a_scode
    
    bb = NaN(size(aa,1),12);
    nanidx = [];
    for k = 1:length(aa)
        tStart = tic;
        aSO2 = stnSO2(stnSO2(:,1)==aa(k,1)&stnSO2(:,7)==aa(k,2)&stnSO2(:,5)==aa(k,3),8);
        aNO = stnNO(stnNO(:,1)==aa(k,1)&stnNO(:,7)==aa(k,2)&stnNO(:,5)==aa(k,3),8);
        aNO2 = stnNO2(stnNO2(:,1)==aa(k,1)&stnNO2(:,7)==aa(k,2)&stnNO2(:,5)==aa(k,3),8);
        aNOX = stnNOX(stnNOX(:,1)==aa(k,1)&stnNOX(:,7)==aa(k,2)&stnNOX(:,5)==aa(k,3),8);
        aCO = stnCO(stnCO(:,1)==aa(k,1)&stnCO(:,7)==aa(k,2)&stnCO(:,5)==aa(k,3),8);
        aOX = stnOX(stnOX(:,1)==aa(k,1)&stnOX(:,7)==aa(k,2)&stnOX(:,5)==aa(k,3),8);
        aNMHC = stnNMHC(stnNMHC(:,1)==aa(k,1)&stnNMHC(:,7)==aa(k,2)&stnNMHC(:,5)==aa(k,3),8);
        aCH4 = stnCH4(stnCH4(:,1)==aa(k,1)&stnCH4(:,7)==aa(k,2)&stnCH4(:,5)==aa(k,3),8);
        aTHC = stnTHC(stnTHC(:,1)==aa(k,1)&stnTHC(:,7)==aa(k,2)&stnTHC(:,5)==aa(k,3),8);
        aSPM = stnSPM(stnSPM(:,1)==aa(k,1)&stnSPM(:,7)==aa(k,2)&stnSPM(:,5)==aa(k,3),8);
        aPM25 = stnPM25(stnPM25(:,1)==aa(k,1)&stnPM25(:,7)==aa(k,2)&stnPM25(:,5)==aa(k,3),8);
        aCO2 = stnCO2(stnCO2(:,1)==aa(k,1)&stnCO2(:,7)==aa(k,2)&stnCO2(:,5)==aa(k,3),8);
        if isempty(aSO2)==0; bb(k,1)=aSO2; end
        if isempty(aCO)==0; bb(k,2)=aCO; end
        if isempty(aOX)==0; bb(k,3)=aOX; end
        if isempty(aNO2)==0; bb(k,4)=aNO2; end
        if isempty(aSPM)==0; bb(k,5)=aSPM; end
        if isempty(aPM25)==0; bb(k,6)=aPM25; end
        if isempty(aNO)==0; bb(k,7)=aNO; end
        if isempty(aNOX)==0; bb(k,8)=aNOX; end
        if isempty(aNMHC)==0; bb(k,9)=aNMHC; end
        if isempty(aCH4)==0; bb(k,10)=aCH4; end
        if isempty(aTHC)==0; bb(k,11)=aTHC; end
        if isempty(aCO2)==0; bb(k,12)=aCO2; end
        
        bb_temp = bb(k,:);
        bb_temp(bb_temp>=9997)=NaN;
        bb(k,:)=bb_temp;
        
        nanidx_temp = sum(isnan(bb_temp));
        if nanidx_temp == 12
            nanidx = [nanidx; k];
        end
        tElapsed = toc(tStart);
        disp([num2str(yr),'_',num2str(aa(k,1),'%03i'),'_',num2str(aa(k,2),'%02i'),' --- ',num2str(tElapsed),' sec'])
    end
    ndata = [aa(:,1),yr.*ones(size(aa,1),1),mm,dd,aa(:,2),bb,aa(:,3)];
    ndata(nanidx,:)=[];
    save([path_data,'Station/Station_JP/stn_code_data/stn_code_data_all_',num2str(yr)],'ndata','header_ndata')
end % yr

toc
